src/graph/expression_graph.h

```diff
@@ -112,97 +123,167 @@ class Tensors {
 
 typedef std::map<Type, Ptr<Parameters>> ElementTypeParamsMap; // keep it sorted, hence map not unordered map
 
+/**
+ * @brief Main implementation of a computation graph
+ *
+ *  Keeps a record of data (tensors) & all operations
+ *  Each operation in a computation graph is a Node.
+ *  Each Node defines its forward and backward steps.
+ */
 class ExpressionGraph : public std::enable_shared_from_this<ExpressionGraph> {
-  size_t count_{0};
+  size_t count_{0};  /**< @brief counter for nodes in the graph; hold current node index */
 
-  std::unordered_set<Expr> topNodes_; // current set of roots. In the end, all but one must have been consumed.
+  std::unordered_set<Expr> topNodes_; /**< @brief current set of roots. In the end, all but one must have been consumed */
 
 protected:  // (these are protected, not private, for ONNX exporting)
-  std::list<Expr> nodesForward_;
-  std::list<Expr> nodesBackward_;
+  std::list<Expr> nodesForward_;  /**< @brief contains all nodes used for forward() */
+  std::list<Expr> nodesBackward_;  /**< @brief contains trainable nodes used for backward() */
 
-  // Holds memory and expressions that correspond to temporary expressions.
-  // This gets cleared before a new graph is built.
+  /**
+   * @brief a shared pointer to the tensor objects in the graph
+   *
+   * Holds memory and nodes that corresponds to tensors in a graph.
+   * Since operations will result in new tensors, this attribute is used
+   * to allocate memory for new tensors during forward() and backward().
+   * This gets cleared before a new graph is built
+   */
   Ptr<Tensors> tensors_;
 private:
 
   std::unordered_map<size_t, std::vector<Expr>> memoized_;
 
-  Type defaultElementType_{Type::float32}; // Type used for storing parameters, currently all parameters have to have the same type
+  Type defaultElementType_{Type::float32}; /**< Type used for storing parameters, currently all parameters have to have the same type */
 
-  bool inferenceOnly_{false};
+  bool inferenceOnly_{false}; /**< a flag holds whether the graph is used for inference only */
 
-  bool checkpointing_{false}; // use gradient checkpointing if true
+  bool checkpointing_{false}; /**< use gradient checkpointing if true */
 
-  bool reloaded_{false};
+  bool reloaded_{false}; /**< a flag holds whether the graph is reloaded: reloaded is true if the graph loads parameters by load() function. */
 
-  bool throwNaN_{false};
+  bool throwNaN_{false}; /**< a flag holds whether the graph throws a NaN exception */
```

Comment:
As above, consider using simple comments for private attributes and align them vertically.